### PR TITLE
feat: EELS memory

### LINKED ISSUE #296: feat: implement the memory.py
## What

A bytearray in python is a mutable byte array, unlike the bytes type whose items are, like cairo segments, immutatble.

We implement it with a dict with one felt per byte.

## How

- [ ] create the type in `ethereum_types.bytes`
- [ ] use an underlying Dict structure with one felt per byte for the bytearray

```
BytearrayStruct {
    dict_ptr_start
    dict_ptr
}

Bytearray = {
    value: BytearrayStruct*
}
```
- [ ] update args_gen.py, remove the `bytearray` from `if arg_type in (Bytes, bytes, bytearray, str):` and catch at the dict/set level
- [ ] build a dict from the bytearray with one byte per value `dict(enumerate(data))`
- [ ] in `serde.py`, update the `if get_origin(python_cls) in (Mapping, abc.Mapping, set):` to catch the bytearray here
- [ ] if bytearray, build the bytearray from the final dict  `bytearray(dict(enumerate(d)).values())`

- [ ] implement the function from `cancun.vm.memory.py`

### PR DESCRIPTION
**Pull request by enitrat**
Closes #296

### CODE REVIEW

File: cairo/tests/utils/args_gen.py [hunk 1]
```diff
@@ -271,13 +281,19 @@ def _gen_arg(
         segments.load_data(struct_ptr, [instances_ptr, len(arg)])
         return struct_ptr
 
-    if arg_type_origin in (dict, ChainMap, abc.Mapping, set):
+    if arg_type_origin in (dict, ChainMap, abc.Mapping, set) or arg_type is bytearray:
```
Comment (obatirou) [inline cairo/tests/utils/args_gen.py:284]: Is this necessary ?
`arg_type_origin = get_origin(arg_type) or arg_type`
`get_origin(bytearray) == None` but `arg_type == bytearray`

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        return ();
+    }
+
+    // @notice Internal function to read bytes from a buffer with zero padding.
+    // @param data_len Length of the buffer.
+    // @param data Pointer to the buffer data.
+    // @param start_position Starting position to read from.
+    // @param size Number of bytes to read.
+    // @param output Pointer to write output bytes to.
+    func _buffer_read_internal{range_check_ptr}(
```
Comment (obatirou) [inline cairo/ethereum/cancun/vm/memory.cairo:184]:
```suggestion
    func _buffer_read{range_check_ptr}(
```
Will be called as `Internals._buffer_read`

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    return result;
+}
+
+namespace Internals {
+    // @notice Internal function to write bytes to memory.
+    // @param start_position Starting position to write at.
+    // @param data Pointer to the bytes data.
+    // @param len Length of bytes to write.
+    func _write_bytes{dict_ptr: DictAccess*}(start_position: felt, data: felt*, len: felt) {
+        alloc_locals;
```
Comment (obatirou) [inline cairo/ethereum/cancun/vm/memory.cairo:122]: Why ?

File: cairo/ethereum/cancun/vm/exceptions.cairo [hunk 4]
```diff
@@ -0,0 +1,9 @@
+from ethereum_types.bytes import BytesStruct
+
+struct StackUnderflowError {
+    value: BytesStruct*,
+}
+
+struct StackOverflowError {
+    value: BytesStruct*,
+}
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/exceptions.cairo:9]:
```suggestion
from ethereum_types.bytes import Bytes

struct StackUnderflowError {
    value: Bytes,
}

struct StackOverflowError {
    value: Bytes,
}
```

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+struct Bytearray {
+    value: BytearrayStruct*,
+}
+
+struct Bytes1DictAccess {
+    key: felt,
+    prev_value: Bytes1,
+    new_value: Bytes1,
+}
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:32]: maybe move this in `bytes.cairo`?

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+// @dev assumption: start_position < 2**128
+// @dev assumption: value.len < 2**128
+// @param memory The pointer to the bytearray.
+// @param start_position Starting position to write at.
+// @param value Bytes to write.
+func memory_write{range_check_ptr, memory: Bytearray}(start_position: U256, value: Bytes) {
+    alloc_locals;
+    let bytes_len = value.value.len;
+    let bytes_data = value.value.data;
+    let start_position_felt = start_position.value.low;
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:44]: what do we do with the `.high`?

Comment (enitrat) [reply cairo/ethereum/cancun/vm/memory.cairo:44]: assert it's 0

File: cairo/pyproject.toml [hunk 7]
```diff
@@ -9,6 +9,7 @@ dependencies = [
   "cairo-lang>=0.13.2",
   "ethereum",
   "marshmallow-dataclass>=8.6.1",
+  "polars>=1.18.0",
```
Comment (ClementWalter) [inline cairo/pyproject.toml:12]:
```suggestion
```

it's already in the dev requirements

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    with_attr error_message("memory_write: start_position > 2**128") {
+        assert start_position.value.high = 0;
+    }
+    let new_len = start_position_felt + bytes_len;
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:30]: don't define here if it's not used right here but below

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    let new_dict_ptr = cast(dict_ptr, Bytes1DictAccess*);
+
+    // Update length if we wrote beyond current length
+    tempvar current_len = memory.value.len;
+    let is_new_le_current = is_le(new_len, current_len);
+    if (is_new_le_current != TRUE) {
+        tempvar final_len = new_len;
+    } else {
+        tempvar final_len = current_len;
+    }
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:45]: just `let len = max(memory.value.len, start_position.value.low + value.value.len)`

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    // @notice Internal function to write bytes to memory.
+    // @param start_position Starting position to write at.
+    // @param data Pointer to the bytes data.
+    // @param len Length of bytes to write.
+    func _write_bytes{dict_ptr: DictAccess*}(start_position: felt, data: felt*, len: felt) {
+        if (len == 0) {
+            return ();
+        }
+
+        tempvar start_position = start_position;
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:117]:
```suggestion
        tempvar index = start_position;
```

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        tempvar data = data;
+        tempvar len = len;
+        tempvar dict_ptr = dict_ptr;
+
+        body:
+        let start_position = [ap - 4];
+        let data = cast([ap - 3], felt*);
+        let len = [ap - 2];
+        let dict_ptr = cast([ap - 1], DictAccess*);
+        dict_write(start_position, [data]);
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:127]: you just need to iterate over `len` and `dict_ptr`, doing backwards, ie 

```
let len = [ap - 2];
let dict_ptr = cast([ap - 1], DictAccess*);
let start_position = [fp - 5];
let data = cast([fp - 4], felt*);

dict_write(start_position + len, data[len])

tempvar len = len - 1;
tempvar dict_ptr = dict_ptr;
```

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+from starkware.cairo.lang.compiler.lib.registers import get_fp_and_pc
+from starkware.cairo.common.math import assert_le, assert_lt
+from starkware.cairo.common.math_cmp import is_le, is_not_zero
+
+from ethereum_types.bytes import Bytes, BytesStruct, Bytearray, BytearrayStruct, Bytes1DictAccess
+from ethereum_types.numeric import U256
+
+// @notice Write bytes to memory at a given position.
+// @dev assumption: start_position < 2**128
+// @dev assumption: value.len < 2**128
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:18]: I think that assumption in the context of cairo means that it's not checked if it's true or false. also I don't see where there is any constraint on `value.len`

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+
+    tempvar memory = Bytearray(
+        new BytearrayStruct(memory.value.dict_ptr_start, new_dict_ptr, final_len)
+    );
+    return ();
+}
+
+// @notice Read bytes from memory.
+// @dev assumption: start_position < 2**128
+// @dev assumption: size < 2**128
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:55]: see above

File: cairo/ethereum/cancun/vm/memory.cairo [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        assert start_position.value.high = 0;
+        assert size.value.high = 0;
+    }
+
+    Internals._buffer_read(buffer_len, buffer_data, start_position_felt, size_felt, output);
+    tempvar result = Bytes(new BytesStruct(output, size_felt));
+    return result;
+}
+
+namespace Internals {
```
Comment (ClementWalter) [inline cairo/ethereum/cancun/vm/memory.cairo:107]: let's not use namespace as
1. it doesn't exist in python, function prefixed with `_` are automatically discarded from `*` import
2. it doesn't work with the current test runner, see #220

File: cairo/ethereum/cancun/vm/memory.cairo (hunk 12 above)
Comment (enitrat) [reply cairo/ethereum/cancun/vm/memory.cairo:18]: we don't want start_position + value.len to overflow a uint128

**Review by ClementWalter — APPROVED** [review]

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**